src/common/meta/src/kv_backend/chroot.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use crate::kv_backend::txn::{Txn, TxnOp, TxnOpResponse, TxnResponse};
+use crate::kv_backend::{KvBackend, KvBackendRef, TxnService};
+use crate::rpc::store::{
+    BatchDeleteRequest, BatchDeleteResponse, BatchGetRequest, BatchGetResponse, BatchPutRequest,
+    BatchPutResponse, CompareAndPutRequest, CompareAndPutResponse, DeleteRangeRequest,
+    DeleteRangeResponse, PutRequest, PutResponse, RangeRequest, RangeResponse,
+};
+use crate::rpc::KeyValue;
+
+pub struct ChrootKvBackend<B> {
+    root: Vec<u8>,
+    inner: B,
+}
+
+impl<B: KvBackend> ChrootKvBackend<B> {
+    pub fn new<R>(root: Vec<u8>, inner: B) -> KvBackendRef {
+        debug_assert!(!root.is_empty());
+        Arc::new(ChrootKvBackend { root, inner })
```

Comment:
I'm not a type expert so I need some help here:

```
   Compiling common-meta v0.4.3 (/Users/tison/Brittani/greptimedb/src/common/meta)
error[E0271]: type mismatch resolving `<ChrootKvBackend<B> as TxnService>::Error == Error`
  --> src/common/meta/src/kv_backend/chroot.rs:35:9
   |
35 |         Arc::new(ChrootKvBackend { root, inner })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<ChrootKvBackend<B> as TxnService>::Error == Error`
   |
note: expected this to be `error::Error`
  --> src/common/meta/src/kv_backend/chroot.rs:41:18
   |
41 |     type Error = B::Error;
   |                  ^^^^^^^^
   = note:         expected enum `error::Error`
           found associated type `<B as TxnService>::Error`
   = help: consider constraining the associated type `<B as TxnService>::Error` to `error::Error`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: required for the cast from `Arc<ChrootKvBackend<B>>` to `Arc<(dyn KvBackend<Error = error::Error> + std::marker::Send + Sync + 'static)>`

error[E0271]: type mismatch resolving `<ChrootKvBackend<B> as TxnService>::Error == Error`
  --> src/common/meta/src/kv_backend/chroot.rs:35:9
   |
35 |         Arc::new(ChrootKvBackend { root, inner })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<ChrootKvBackend<B> as TxnService>::Error == Error`
   |
note: expected this to be `error::Error`
  --> src/common/meta/src/kv_backend/chroot.rs:41:18
   |
41 |     type Error = B::Error;
   |                  ^^^^^^^^
   = note:         expected enum `error::Error`
           found associated type `<B as txn::TxnService>::Error`
   = help: consider constraining the associated type `<B as txn::TxnService>::Error` to `error::Error`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: required for the cast from `std::sync::Arc<ChrootKvBackend<B>>` to `std::sync::Arc<(dyn kv_backend::KvBackend<Error = error::Error> + std::marker::Send + Sync + 'static)>`

For more information about this error, try `rustc --explain E0271`.
error: could not compile `common-meta` (lib test) due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `common-meta` (lib) due to previous error
```

Reply:
I can learn from `CachedMetaKvBackend` anyway.

Reply:
Fixed in d02781c18e03ec8fc000efe107586fed451a6f54.

Reply:
Cool, I'm working on this tricky type. Anyway, `KvBackendRef` is also good.

Reply:
Yes. I ever has a local version to use generic to pass the compiler check, but it's squashed 🤣

Reply:
Somehow it looks like `impl <B: KvBackend<Error=Error>> for ...` and you need to `impl KvBackend for KvBackendRef` which is another taste that I cannot judge if it's good or not.

Reply:
> you need to `impl KvBackend for KvBackendRef`

Or let `EtcdStore`'s constructor returns bare `Self` instead of `KvBackend`. Anyway, too many subtle things.